domain/network/state/space.go

```diff
@@ -0,0 +1,323 @@
+// Copyright 2023 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package state
+
+import (
+	"context"
+	"database/sql"
+	"fmt"
+
+	"github.com/canonical/sqlair"
+	"github.com/juju/errors"
+	"github.com/juju/utils/v3"
+
+	coreDB "github.com/juju/juju/core/database"
+	"github.com/juju/juju/core/network"
+	"github.com/juju/juju/domain"
+	"github.com/juju/juju/internal/database"
+)
+
+// State represents a type for interacting with the underlying state.
+type State struct {
+	*domain.StateBase
+}
+
+// NewState returns a new State for interacting with the underlying state.
+func NewState(factory coreDB.TxnRunnerFactory) *State {
+	return &State{
+		StateBase: domain.NewStateBase(factory),
+	}
+}
+
+// AddSpace creates and returns a new space.
+func (st *State) AddSpace(
+	ctx context.Context,
+	uuid utils.UUID,
+	name string,
+	providerID network.Id,
+	subnetIDs []string,
+) error {
+	db, err := st.DB()
+	if err != nil {
+		return errors.Trace(err)
+	}
+
+	insertSpaceStmt := `
+INSERT INTO space (uuid, name)
+VALUES (?, ?)`
+	insertProviderStmt := `
+INSERT INTO provider_space (provider_id, space_uuid)
+VALUES (?, ?)`
+	checkSubnetFanLocalUnderlay := `
+SELECT subnet.cidr,subnet_type.is_space_settable
+FROM   subnet_type
+JOIN   subnet
+ON     subnet.subnet_type_id = subnet_type.id
+WHERE  subnet.uuid = ?`
+	findFanSubnetsBinds, findFanSubnetsVals := database.SliceToPlaceholder(subnetIDs)
+	findFanSubnetsStmt := fmt.Sprintf(`
+SELECT subject_subnet_uuid
+FROM   subnet_association
+WHERE  associated_subnet_uuid IN (%s)`, findFanSubnetsBinds)
```

Comment:
I had thought about this, but since today I didn't see any new type of association I thought it was just noise. But indeed it is more explicit for the future. Done in e327ee072bd0f545f10983f8b52de35a238df2c7